### ISSUE TITLE
Proposal: invoke and await nested onbeforeremoves

### ISSUE THREAD

**Issue by barneycarroll** [MithrilJS/mithril.js#1182]
### Description:

A sequential reading of the rewrite docs fills me with joy. [The first time we encounter 'lifecycle hooks', it turns out we get very fine grain control, and we get it for any type of node!](https://github.com/lhorie/mithril.js/blob/rewrite/docs/hyperscript.md#lifecycle-methods)

[But then it turns out that for `onbeforeremove`](https://github.com/lhorie/mithril.js/blob/rewrite/docs/lifecycle-methods.md#onbeforeremove), "This hook is only called on the DOM element that loses its `parentNode`, but it does not get called in its child elements."

This is irritating because whereas the documentation up to this point has encouraged the idea of allocating lifecycle imperative code to whichever nodes it is most relevant to, this sudden exception forces the author to violate that principle and introduce their own boilerplate to account for something which could be done procedurally.
### Steps to Reproduce:

[Here's a bin with a component featuring animations](http://jsbin.com/keruga/edit?js,output). Click 'Edit' and then click 'Done', then dismiss the modal.
### Expected:

On dismissal, I expected the modal to execute animations and wait for their resolution before being removed from the DOM.

The beauty of this code is that I don't have to push all my animation concerns into big centralised functions: the entry and exit animations are properties of the elements that they relate to. I can adjust them in place along with other concerns such as static styles, content, etc. The dom node is provided to me directly and the entry and exit animations are declared adjacently (useful for debugging!).
### Actual:

Nothing happens, because technically these DOM nodes aren't being removed — one of their ancestors (the component root) is.

[Here's the same thing, refactored to work around the shortcomings of `onbeforeremove`](http://jsbin.com/bewudub/edit?js,output).

The ugliness of this code is that despite the implied potential of universal lifecycle methods, what is arguably the most common case for non-component lifecycle hooks — animation in and out — is impossible: so we revert to the dreaded 'fat controller' anti-pattern, where conceptually distinct concerns that we would otherwise expect to be grouped into their local sphere of reference are instead shoehorned into a top-level function that needs to do extra legwork to get the right references and set up the right conditions for async resolution, introducing indirection and brittleness. This looks silly, because I've split my entry and exit animation functions, which in my mind are intrinsically inter-linked, across large portions of orthogonal code. This pattern mixing feels wrong, and as a result I'm tempted to keep all animation code for a given component in its central lifecycle methods. At this point, the whole idea of per-element lifecycle hooks feels hollow.

Other minor points to consider:
- Deep querying the DOM from within component lifecycle methods feels like an anti-pattern. When this is necessary in Mithril v0, one at least benefits from the neat pattern of being able to bind element references to stateful props via `config`<sup>1</sup>. This is considerably more verbose with Mithril v1's lifecycle hooks.
- It's tempting to read some kind of conceptual purity into the present implementation: the nodes in question aren't technically getting removed - only one node is getting removed; but that conception is inconsistent with eg `oncreate`, which technically creates leaf nodes first and appends up the tree — and yet `oncreate`s trigger in source order, and allows leaf nodes in fresh trees to query their painted properties after their ancestor has been appended to the DOM. So the justification is internally inconsistent.
1. The `prop` + `config` centralised DOM reference pattern:

``` js
{
  controller : function(){
    this.el1= m.prop()
    this.el2 = m.prop()
  },

  view : ( { el1, el2 } ) => [
    m( '#el1', { config : el1 } ),
    m( '#el2', { config : el2 } )
  ]
}
```

**Comment by barneycarroll** [comment]
I make it sound like this caveat is sneakily hidden with the malicious purpose of disappointing me, but actually the very first mention of `unbeforeremove` in the table of lifecycle events specifically says:

> This method is only triggered on the element that is detached from its parent DOM element, but not on its child elements

Please read the OP with an indignant outrage filter and colour me chastised :sweat_smile:

**Comment by lhorie** [comment]
Theoretically, it would be nice if onbeforeremove could trigger in any vnode, but in reality, the edge cases and the difficulty of implementation make my brain hurt. For example, with those semantics, if a vnode is removed and one of its descendants has a onbeforeremove, then the parent will not be removed until the callback is called, and it seems somewhat unintuitive imho.

**Comment by barneycarroll** [comment]
Why is that unintuitive? You simply flatten a list of all subsequent methods, reduce them to a count that needs to be resolved, and wait for them all to execute. There's no new structures involved, conceptually. To put it in v0 terms: flatten the subtree, extract all onbefores, and m.sync those before the current node can be removed. 

Am I missing an order of execution problem?

Well, I'm definitely missing *something* because as much as #1184 works on paper (introduces new tests that I think cover the problem, both pass), when I use that build in my problem case above, the component is still removed immediately.

But if you look at the patch in the render function, we already have a count check to wait for and I'm incrementing it synchronously before we run the resolution condition.

We can do the whole condition determination thing synchronously, right? I mean, I can do that proceduraly, without thinking, in application code in the second jsbin there in the OP (the only difficulty is in maintaining my DOM references). I parse the static subtree, extract the pertinent resolution conditions, and wait for them all to resolve.

What have I missed?

**Comment by lhorie** [comment]
> Why is that unintuitive? 

For example, let's say you have a bunch of widgets, one of which is a list of things, and that on deletion a list item slides out. If onbeforeremove is triggered upon ancestor removal, then the transition would fire for all list items on route change thus preventing an immediate route change

On the implementation side, the thing that makes my head hurt is that while onbeforeremove's callback is not called, the vdom tree is out of sync with the DOM, and I'm not entirely sure what would happen under the various diff scenarios if another redraw happened while that state if the entire tree can be out of sync (rather than just the element that has the hook)

**Comment by barneycarroll** [comment]
I don't see the qualitative difference between these 2 problems:

1. During view reconciliation, a node is tagged for removal, and (maybe) the lifecycle is frozen while arbitrary author code resolves, at which point (certainly) the next view snapshot's lifecycle begins
2. During view reconciliation, an entire subtree of nodes is tagged for removal, at which point between 0 and all of them can opt to freeze the lifecycle until all such conditions are resolved.

The possibility of a `done` not being called or a state which supposedly exists *now* not being represented because the last state's teardown is async… These problems are already here. If you think these are inherently bad things, avoid this API; if you understand and value the bargain being made… The ability to describe subtle nested animation rules is not making that integral bargain qualitatively worse. Credibly, the difference is a quantitative thing based on designer judgment: am I using too much animation? Is this pretty effect worth the delay in persistent new data to screen?

**Comment by leeoniya** [comment]
in full agreement with @barneycarroll. once you've given the user direct control over vdom state you always risk that the user will never call done() to either unfreeze a locked tree or let the lib finish a partial transaction. i've gone through all this reasoning myself when giving users the option to return Promises from `will*` hooks.

without this though, you end up forcing users to jump through a lot of hoops for trivial things like removal/unmounting/reordering animations.

**Comment by spacejack** [comment]
As a user, yes, it would be nice to have onbeforeremove to fire when parent components are removed. If it's too problematic, an example workaround in the docs would be nice to have in that case.

**Comment by mindeavor** [comment]
I likely don't fully understand the problem, but what if...

- `onbeforeremove` was recursive
- Returning `false` from its callback meant you stop the event from propagating downwards
- Changing route is like "returning false" by default?

**Comment by isiahmeadows** [comment]
@mindeavor I think the biggest problem is that `m.redraw()` in the rewrite is fully synchronous. And async animations don't work well with that.

**Comment by lhorie** [comment]
I was thinking last night that maybe I could have nested onbeforeremoves be triggered, but for DOM removals to happen on a first-come-first-serve basis

for example, if parent schedules removal after 1 second, and child schedules removal after 2 seconds, parent is removed at 1 second

another example, if child is scheduled to be removed after 1 second and route changes, root is removed immediately because it's scheduled for synchronous dom element removal

so in the case of the hamburger menu that needs to close before route change, you can just add a delay in the root onbeforeremove to give the animation enough time to run, and if you don't want the delay, remove the onbeforeremove

The only issue is that you can't selectively tell children not to fire their onbeforeremove, potentially unless you do some shenanigans with setting flags yourself

**Comment by barneycarroll** [comment]
I wrote a little playground to explore the various edge cases around this, which might be helpful for trial and error: https://jsbin.com/zenoyi/edit?js,output

On reflection @lhorie I think your suggestion above is probably undesirable in the grand scheme of things, given the overall picture of what `onbeforeremove` can and cannot do. In conclusion I think the best path forward is to leave the current behaviour as it is, and write up some guidance as to what authors can reasonably expect from it.

The current behaviour is on-par with what's offered by Snabbdom and others, where the documented use case scenario is individual items being asynchronously removed from a dynamic list (in which case I probably wouldn't want every item on the list to perform outward animation when the containing component disappears — that's a semantically different thing in user terms). @leeoniya points here and elsewhere to the fact that a deeper awareness of deferred DOM patching is a bigger challenge than other competitive virtual DOM libs have achieved (that would be best served by a more holistic declarative API), and as you've said yourself @lhorie there *are* ambiguities of intent here which would be difficult to shoehorn into the current API, let alone implement.

***

…Here's my abortive attempt at implementing the compromise [you're talking about above](https://github.com/lhorie/mithril.js/issues/1182#issuecomment-237249299) in user-land via a wrapper function for `onbeforeremove`: https://jsbin.com/zimiga/edit?js,output. It's naive, and broken, but it gives some indication of the complexity of assumptions necessary for implementation.

This exercise gave me a good deal of sympathy for #1086 (the new vnodes are so complex!) and #1229 (APIs with higher order flow control implications — event handlers' `e.redraw`, stream's `HALT`, `onbeforemove`'s `done` are painfully inconsistent and extremely hard to extend / intercept). All in all I really miss the redraw strategy API! 

Having said that, it's now clear to me that the kind of behaviour I was originally calling for *can* be implemented in plugin land (I just need to patch `m.render`, hehe >:D), and *shouldn't* be expected from a baked-in lifecycle hook. I'll get plugging away at Exitable rewrite and report back when I've got something usable :)

**Comment by isiahmeadows** [comment]
@barneycarroll I'll note that this *might* become possible if `m.redraw` becomes async (there's recent talk of this in #1166). In this event, I'd rather not nest them, though.

**Comment by isiahmeadows** [comment]
Resurrecting this bug based on [previous discussion in a separate, borderline-duplicate bug](https://github.com/MithrilJS/mithril.js/issues/2090#issuecomment-366705074). Retargeting to v2.

**Comment by isiahmeadows** [comment]
BTW, [here's a copy of my proposal from over in that bug](https://github.com/MithrilJS/mithril.js/issues/2090#issuecomment-366721169):

-----

I've had a couple years to study highly async stuff further (including async DOM manipulation, which I've studied for quite a while, and other batched async updates), and here's what I'd reasonably expect:

1. All hooks are invoked at once, starting from the child and up to the parent.
1. Nothing is removed from a parent until its `onbeforeremove` and all its children's all resolve.
1. Once everything has been awaited (or if nothing needed awaited), then remove the tree.
1. This works holistically. If a child resolves before its parent, the child waits until the parent resolves before it's properly detached.
1. `onremove` hooks are fired in a separate pass after all `onbeforeremove` hooks are passed.
1. After those are all fired, the parent and all its children are unlinked. (The [`TreeWalker` API](https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker) could simplify this for us *drastically*.
1. Any future `m.render` calls on a tree would be blocked on existing `onbeforeremove` hooks. (This would also resolve [this bug](https://github.com/MithrilJS/mithril.js/issues/1881), which we need to do regardless of the outcome of this one.)

We could also add an extra options argument to `m.redraw` ([related](https://github.com/MithrilJS/mithril.js/issues/2074)) to accept a `resolve` and `reject` argument, called on successful and failed application of a new tree. Although those wouldn't be applicable to `m.route`, it'd be applicable elsewhere (like in tests, where we could actually check for scheduled timings). We could also use those to provide hooks to make `m.redraw()` (async version) return a promise that resolves when finished and have `m.redraw.sync()` accept an extra options object to also expose this facility.

-----

Needless to say, async isn't exactly easy nor simple, but that's mostly at the theoretical level. The implementation is much harder to *plan* than it is to *write* - it's hard to get *right*, but it's not super *verbose* to get right, at least not significantly more than what exists in this instance.

If we agree to this, I'll try to commit to it personally what I can.

**Comment by barneycarroll** [comment]
`onbeforeremove` (OBR) currently means "this node will be removed _but its immediate context will persist_". That last clause feels like an obtuse burden the first time it surprises you, but it's a useful guarantee and removing it opens up some awkward ambiguities.

You might start writing a tree with the expectation that descendant nodes' OBRs would be executed and allowed to hold up branch removal when an ancestor node is scheduled for removal, but in many other scenarios it's completely undesirable and difficult to differentiate from the distinct clause of the present mechanism.

Consider a leaderboard component consisting of a dynamic ordered list where items are expected to move about in relation to each other, and appear and disappear within the lifecycle of the leaderboard. `oncreate` means the entry is a new contender in the leaderboard. OBR means the contender has dropped off the leaderboard. These have specific meanings and any action within those hooks is liable to be intrinsically associated with those semantics. So in this context, the idea that removing the leaderboard from the UI should trigger the UI for all contenders leaving the board is undesirable, and if that were default behaviour, writing extra logic to determine when that is and isn't the case isn't intuitive.

On a more generic level, imagine an app making use of a consummate suite of generic UI components with incoming and outgoing animations: modals that slide down and back up out of view; form sub-sections that collapse on completion; media view panels that expand and shrink from a point of origin, a side bar that slides in and out from the left. Now consider the user navigating away from a page with a combination of such elements in view - back to the home page, or to their profile, etc. Do we expect all of these elements to attempt to simultaneously animate out? That sounds like a designer's nightmare.

In general the idea of procedurally triggering OBRs for all descendants of a removed ancestor violates the principle of higher order authority and isolated concerns. You'd have to introduce extra code to avoid the scenarios where it wasn't wanted, and in situations where you're making use of third party components that may be extra difficult.

By contrast, writing code to specifically instruct descendant nodes to animate out when on ancestor OBR is common stuff for UI developers. The relationships are explicit, the individual animations can be coordinated and sequenced according to holistic concerns, etc. It's not credible, coming from a perspective that's concerned with the practicalities of such things, to have Mithril do that sort of stuff 'automatically'.

**Comment by tivac** [comment]
>It's not credible, coming from a perspective that's concerned with the practicalities of such things, to have Mithril do that sort of stuff 'automatically'.

Interesting to note the complete 180 in @barneycarroll's thinking on this over the intervening months.

I'm :-1: x :100: on this, it seems like a huge source of unintended consequences. The current behavior is a reasonable compromise w/ an eye towards performance and given how few times it's come up since mithril hit `1.x` I consider the anecdata to agree with me.

**Comment by isiahmeadows** [comment]
@tivac I've 180'd too, but to be fair, I've actually *designed* an implementation for such a feature elsewhere, and so I would know how to implement it. I also see the use for it, in that the few who've hit this and pointed it out have been

> The current behavior is a reasonable compromise w/ an eye towards performance

And a naïve version of what we do would not substantially affect performance absent child `onbeforeremove` hooks, particularly if we [go with this alternate route](https://github.com/MithrilJS/mithril.js/issues/2090#issuecomment-366726467), which is also much simpler to implement. (I edited the original re-opening comment to add this part in.)

-----

If we choose to decline this, we *really* need to better word in the docs what *really* happens.

> This hook is only called on the DOM element that loses its parentNode, but it does not get called in its child elements.

This is *technically* correct, but we don't use that phrasing *anywhere* else. People tend to miss this, and it's written in a way that emphasizes the DOM, when we abstract and gloss over that pretty much everywhere else in the docs. It also is what has run people into problems with routing animations. It could probably be better worded as this:

> This hook is only called on the top-most DOM element to be removed, *not* any child elements. If you use [Mithril's router](https://mithril.js.org/route.html), note that because routing changes the component at the top of the tree, this is *only* called for the top-level component.

I do feel that the majority use case would be resolved if [we just made routes like `render`-only functions](https://github.com/MithrilJS/mithril.js/issues/2073#issuecomment-358485564) - a common layout, for obvious reasons, *wouldn't* be removed and re-created in that instance, allowing `onbeforeremove` in child nodes to potentially be invoked as per usual.

**Comment by barneycarroll** [comment]
[I wrote a component to allow explicit `onbeforeremove` dependency graphs.](https://flems.io/#0=N4IgzgpgNhDGAuEAmIBcIBCUD2sDWEATgHQBWYIANCAGYCWMFqA2qAHYCGAthGpjviJkK1WNjaIJfeIQCeAAmAAdNvPldsSAK4xiEAB4AHbIXhh5AXnlZceFQF8VsDvFgALABQQAlMEdsVGi02BDpxeWcoKAB5DAAlDwA3b0UVNUIIeC1CVQAFQmwuOkhiDiiPZjS1eUTS+BlzADJGmrqG4nEAIwgaEyIIDUSIeWaq6taXdq6evozBiGJIqCTiMHgXCEoa7yrKMdb1gHMOtm7ejLnsIZHG-bVao5Oz2YGrhaWVtY2t5KqAXR2bH8gWCoXC9DYSFiCWSylU8gA7m4GMMVjA2Id4G55AAfHHyACEK0e03O-XmuPxtUmhCaLWp9VpTxmF1eQ28gPGiUsrTobC+IWGeNa7gYSAyqmFiWYAAY-ioqhksjkag4FWwgiF4GFVIYiGBivAkr4qokOIR5OtCIdMjyIVD4is+QLYEKqYtkVBxRA2ICqnQaPIVjSwMQoHy8Jzqgz2uG2HhVpkPFabfBiEhClsU5lAcCNaDteEbIJCB4TfCzRa43geWwIAj5ABZDiGKqKzLZVRwrl0evG1Lw8ZKztBu7jCaM0OJXsIsdqAD8c+jbSZ0771aj4-kqCX9w9YolY838nse0H8nEsAyG1Q8j1tMNZ-G4i0hiQN7v+sf+3ElyG-e7ccAyDGMmQ3JdQNDat02gTIIBWDMuGPU8f1OFlyTeMsBy3YcVXyQpigWMpliXABBQhCA4WRiBoAouA8aCzSgLQIDAMtjy3YguBbDwlmhDj5GQqp-H8EBT3YbheHQUi2Dobj4BMYQqBAMQJB9eA+FUtZ5BkuSXBMHlALXBtb25CwAD55H2eiAHJmDWWQYAsbBDA4WA6HgWRUBlABuJBikMKAqNQVTJDMP4bKfcdalFL1D3PbwoqqS9r0Qbcakscz9lqRDSlk+T4MA59XPczzb1lLYAEY-ii8YZA4flzi4cqbPq-kgsQABNDwAFoAFYZQAUm8SL5Fayj2o2bqZRGmr9nEsd6CiW8bPOBFzSQMBIrHYhdsg4g+Vqk9ErbeFSReClTMyqzzzrBt8KKSAPAzOtrqK5dcoavTEA8d7ircjyvPkZhKq2OUjrqiawCalq2rADqIGmkatnGhr4amjwBuGmy5vPNQFrx6olqgFa1o2raIeqXbiH27AtHgSn7G8E4IWKbErBeiB5tzMTKAkng+CwbJOiU0RxDCzTxG0oXCE6QzTRndKsIs4HrI8GzDFGgADMdizwLh5GKeQOAiQpjDrCRLTcFxjaibAEXMWQ6ctbB5CQaAqONpsPLcQgGHkNhNAgGzzDHc7WQpDIwGwZjC1UBT5B9MBsmGMA9Xc+hkDd1jXUhBqzHUBqOBtF2bq3IIoCWq2IDoWkXewYh9i1k6AnPWzNa2HXCb102uHN9TzEMAptFdAPXYAEWiRsL0DLEa4te3VB66uFBrueLQMOB6eGMc59r+Rk8MQLeyQGpFZpOhOm3i8LT3i1Yu9VR-IyBAoGopuW7VjXtbHAAVNxhj0FpPAawAgaxGznheV6SJsCQHkH+MoiJhjrUtgnd2QV36EwAJLx2ROYAw3BAqbENvAEOXtwzaSRHQdwiIGBQFNpCDyOpiIKAlO7C07s9SMIxGHXBwx3BwDwJ0bA+h5DdBwAiYg8h-7DAoSAwO7t5DW3MIHA+6c6CIIQfQjIABHLQtcBgDzDoGDy5hF5Bg8qbHQp9rZDC2FfEBliMysTYK1RKYj6ZIPkHgQODZTERE9I-N2DcxwAGVXYIn4cRL2PcGqn0OHQa4liTbGVLu7MAV5L7DH8aFAejdzzNzPF-Du8gu7jn-jbVJRsh5XDoO7U+CIfZexstWGyxtrRaB4BIKRAB1ABqhIlKI4NcE26Tc7vgkHcBRyC3CwOGOHDC1wwCzOsUomOp9XzV2NoKNYJgtiDIRJRQwJDDat3HA1bA69QG2F7v3CQWw4l3gKNORRWIbazOwDWDg5gTYtIjG0i+DiFgfyKW3dWQ9g6dxULZAAxC2DuVRuKyG6DcMu1R6J6y2MAeQAB6HF8gsGzzwYbcwCcUV-mQNTMcxlbzVmuni+QAAeIMkDuIJNgGIjyAk1C2W4nyYgWI+SHG2oTcYGKwFYurKeNQDK55sCpaK8c9FdLySFQAYTNuIdSWwGXU0FeYRpUQA4QCzgnVmyz5BALWLuccNlQlZBoDQEVNqxUeExYoKVOr8UONttHBVW4A3oo8Cqlw6rNUWwZri-Feq8EuttY2PoB8HU0AJCKwNW5uXjkzWoTNyFATHkKf8BwvN+ZSRAHyd2+hRYqXFupPgXAuJ0wkM9XAnT1LplbV0+AABRGAXasUK3rLeX6XxEBM2ul-YRSBZD2U8k5XoEhUDEAABx9QGAfNGPVIB+xoD5VySB-IYlQJVAYEVKa2TcJVUaS4bJ6yFWNLYS5bJiHdrOxyEALALvgFuugAAvCAx7iAAGYgOnuvYqtQNkFkUpskudxN7s4ZJ9BMgujSsTXPwFwWDhNP6E1skFcRaaA22T5IYemzBPJ6gsAI-Awj9BnuwummjBAkC3lHQsZZ9tKbPhCNbDEAGgwpBVux1YsyGxWAJCJzjCJuPHVBYGtpoSxPyDkdhjN8nxxSeU80NFW5xW2CIAOiDZ8h0qYjJlONzAbKdHNHWWQzgKIxygKNKDYAbOuJRsUDRbgeAcAzA7HGca1BcR4h4DyAwth0ElT6TEbhx0WSC0GvWhnGPpoDdWWTgbpWJddSGhShAjNpbS3yW8f0itbggN8oVK1KuQB6nTUhmXyvZwwbeOg8gABU8hKoDSa+V7QlE463iAzKGU8gADUFjOvyAGi3ZrAaUJnPm+OBrpWcsBtq9VsatWIA9T5ER5b450HBSDOiWL8hl7teXpVFIXWesyj60VgbobxDDdGxN07MX0OXYu9127M2HvrfGP4Q7+NHuBvbuB0HSr1YcAo+aVMFhoCQCRP0BjZXoe+x6CtNw9RDBgFQHihJWItAizEFwHFbTJvha4ED8c2XjPzZsgAASpyQgYdPxjZqK9zwNvOs2JczUzSmY5xXC2PLmbwYkaogGrEwVgIBOAC3QEULEfsoDKWyBr9AuP4D48Jzi4Ihg8DHHJzi1XvsGDM8qsQG3AA2c3Pt1fVso2W8gJbFeST4JEzoPUvqqqlj1d31Atd8F1-rvFRuTeLEKDin3fv8ovf5EHsAzOABMwHbdx4gL7-3SfQxFHlcHkAru+Du-sH8ewQA) I'm planning on polishing it up along with some other view components as a toolkit for composing view concerns in Mithril — I'll post back for posterity when there's a published solution for this.

In the meantime, shall we use this ticket to earmark the documentation improvement suggestion?

**Comment by barneycarroll** [comment]
I'm maintaining a component called [Waiter](https://github.com/barneycarroll/cascading-components/blob/master/src/packages/Waiter/index.mjs) to abstract this functionality as part of my ['cascading components' repo](https://github.com/barneycarroll/cascading-components).

The API works as follows. A higher order Waiters is passed a function to determine the sub-tree, and that function is provided with a `link` reference. Then lower order Waiter components are used to wrap immediate children whose OBR we want to be triggered by, and delay the removal of, the higher order component - these are given the `link` reference as an attribute of the same name. Straw man:
```js
m(Waiter, link =>
  m(Waiter, {link}, X)
)
```
[In this example](https://flems.io/#0=N4Igxg9gdgzhA2BTEAucD4EMAONEBMQAaEAMwEskZUBtUKTAW2TRiUWwDoArakyKABdEQ1CEQAPbBABOggAT5EpTAFd4CpVgCe8gLwA+eeSjZVCwwB0o82-KiIA7vIAKMiI3J4AFDMRx4ADdEfSNgazs7PEEAFXJmCHNvbwBKUPlwm0i7PwDg7xMzQRSI7IBfIkVEHRKs2zKU4hA8JDBBcmhqNAAGFAAmAE4QCvomFhAAWQh8THgePnQhEUExARgFdrAAa315VPS3Dy9ETlyEfNrrSWk5KpV1BUzbQPInFD3gMAALSnw-KAaoVKtkY3gA5JwwfIANTyABGAANgXY4ZhtgBzdyqKD4eQAYlIhORtjhEAkAFoYF9MPgIM48d1GQBGeTdVnyPqIRjyTh9ACsXOJ8mwNPwJnRkSZgrqdmkMHI7WgdgoEgIQsEEGw2TsfO6AFIhUhSAptbqDTLbIIZJhYKRZNz5FabWxMMJvOSzZUPfratrGJgKY5yPhBF9bAMAByBRxC-0Ur6Icjor4KSOBL7q62wBUdGycAAsMHkiEweHJJnJiUEQs1aIVujs3WRCMqT2y0DAfldiHe3mAtMY72A620SEqJgcMgAEonk4IyoDDBkhbYR0hOLWwPXdkzrCvHeRtqlOKGRN5GKcCNbHL7ykQ9xb5NA4cpZIg-IwIMFe-2PIujA4ziHJ4PhnEEIRLm22oDpwa4nJu256KyD7arYMGigAosEQgADJeMIk7gk62aKg4OJgpUYHBLekQNMiFRCt8vz-MiKT3gCTQtIgbS5l0IBMvmKBMkMIwgAwzBiAA6pgCrvvMTQCARKxoNcsiaMoagaHsaSQaUgSYDI8jwCYOxIYB8gTDgKHyH4giqDINhQfILxOPI7yBOk3j7oEnCulaMCcMZUBbEKAD88jeZwTHwH8IhCigkXRbFUA0N0AC6UU-DF-zeEFWxpKx7HIh2XbCO82DvvK6zsZE0CqNgMxlcKlX4UVMrQB+X49s56RObY5CkN4Pl+TIAV5TR2TDYI-mBSZnBaIgbo+QOE0MVAxVQC+dp+J137aUCj7AccvnwPAXmPrYNCcNdeWcPp8CqP4qQZf62DeGAszwAA8gAQgAShNti3mU1gg+tUBrAoH2nb9f27KW2gQz1ulZEocKqOi6LvqUpSYI4MkKEdeAnWdNArlN-nIgAZFTzm+dNo0bptr47VyXXU7TFOM8+LPvmzwRRZ9Q2wYI3aVIEbHWZNIvdkzW1vrtiAc3T6yyzz2185+AvQ2dPmq8I4uS1kaW3qUF5+H8eOpKDD6Q-IFA4rDuweSjtiOFlITC0gUDoqG8gAD7+-IACEwui+icu84rAdB1zRY03TI0BerCv84gKQTS7dMmKrEMhIHdNJf8MfOalaXWbZ9k2IENvg3bFWjfhzu9aUduizIWMWPbJj4LDQ2XFkA3C0ns3BZn9Mzbd0TeO3nfzR4lSz4ttRlJx1TcaRfEAMxMkJ+bDEQowSWgACSUBBScvAKdASliKptxKPcWmkNiPHQP3bZ26QuzndqPlFyIMuEUZQ03VNoCqEBv7-yyslMumVmIiH0HoJCYIX4Q1ImCDmQpoEIJSuleB2U4pZAHrYSuDk9hClIMiUKQpf6oVsKQTg2B3AakEOAkICdGHMIgKw9hd1XgxguvIGhQjbDmUGtdXyHdVDMCEDAQGdgEqiPtt4SRBl0QyOWPIoUE0lHaj6nYFyjgP77lIYtKuFDlGRDjvwpwpi7AiPoX-CejMjGqOuuozRciFHaj0U46xLiAoAPBv4nx8gwbagifIFeB8j7jC+oSeS-Ab7LFWJ0BQDcqpdxdmEUoX04TcA3pwLYiBtAwGHgzeRnBtoYTRF8Ge7CW4yj1qLYQnBEDYUEAFUUDSKokLsEGHEdJfL4HwFhZYeF1giHfL0xA4sZYG0dDIR6wMV623SY6EsMhaSOBsEhHJy4shXWui02WHStEZRqXU2ZTTIiDJ2ZeLWiBxm4XwtMmQsz5n6zmUslZpQ6IAgfPfdST8FAHL7HpARbl9pWGaQQ5KbVbDQBMAqaF+xXaGIWScc5cjdjmQAMqLX+Yip8ENSohHKi1aqpQ6oNW7NCzJrUaVQGju8YQBkdlQBJdSHESAXmCHeBBXJMoh4nDYRVYw1dAnyATsHZaHgoo3xkrAbwYr1HLxSAYuONAxXsLSqq-pkRRWXkts4YOyD7azDwGE82V4rarWJdZFOfh6qNW6lgYQ6xbmGN2B6-w1YshrQaLEsSYwxBEzklfZJSxRAqSkGpO4mlHiQreMjAwyIcGELwRlYJfZJGnOEA0ElyKoAKneAjJGBynL5MKW0XyMB5ToigMLb5rZsGzEeu8QxEBgyshqtkd87gZBdueD23E3R+2RCogQd4Kh4B4EnTkRAtbhD4FnVauZQoKo4nFF2q0j1F2rkWoIJAa6GEbsXQC5EVpdBaqxXdDtEF5B4wJonSpTD3AgSVo+AtJxp24iQvu79tFkQfUEN8A1d7vntJkEO3YwHpbQb8CugguwgP0WRA7T6t7sH3u3WKH2uw514Fw9B6IJ7UOAeWd+2MJrrzWxlFEtazL5YurpU1bwUA6SLwTFAHSwqjWDS444QJH6jh4BDha08UBRPcK-Qo4TJ5MARxLQqTjdJgbsVXiQLib9YBiC3oJJkIbxLjCSYsW+aB4g3AUGwRAHBIikE-fICEAB6OzHBMHg2swm6SslDIMOc25vzwgZBeZ87cM+F9lRBc4K5qLJhEDhcYDZ+QCTv7ZCcx4FzcX0vJdSxGwyWXuRucK156wEWFBTBmPARzsXXPVdmOViGGyACC2AtS7AhFCWEiJSh2iEOSFQnh4C6BgM6Sk74BqlBFKM8U2QpSMGsEicGSAFDKb8JRy186aNQAvJ+bEghvC0jAF4wQnBST4G0G2rIRjez8eAZEUE7XsCVFjOCL4TIevwgRMICQghySzCTDYMAyx3wthczOU6EAwRG21LQ+xwcNv2fwPuRxTjQQhffJUPK6R7Egm8Olm7-jIhgGMtsQV3qSd2GR1t9DVi7BrWpxUR71P5CY4JtjjIeUWf4+yKCRr8A3sM+1GCH6uhFImFUPNjU8hVASdDF4eQ8pFkUT56L7QiRn2Yns46CAmzYPkDhEgR0PwfYBTVyL-n4I4Rhbh2zyIYIpyYGCM++wh4QgzG0MHS3DugaHqcfb0JAfIhhL8fQ0ECWHDE5J6pgVHxvn-kOQ76Dcnjj6HV3YW1ABHR66wPvTVwCgVzrneDQGwFgUHXwEBKBkCecBh5pgnEgIwVzcoumua+2E-xJ5ePeA829KUW8M586ZyTvnd2E+tO6sANPeAFz8cz6CQrrY5+IBZ32fD4pKiDtkOLR9Q52gUcqJd7QZRdjAAX55TPkQt8+xv8Ih-BOwTYG+4iP7AOgdNvkKDpYMgWxP52Bgg4QQA0jb7ChIClghD4wKiSJeZ+52Dd6oTh5+674yAP7hSAFkxW5+jgiv4wg-aAGRAf6A7GTf6-6has4IGk4ICyDah4hgCMhEHyAAE4GoRggACqXwlYXwnA8g+KHgi0ZuEojgyw8gjg7gPsKA8B1B-uUs1BoIL+fgb+K2shkQqIGIWIOI+IpAYAYAzBtgs2BGEodgi2Bh8Isgte2QDBTBbBkQrBahdgaBzBSB9C5cbBKBDu2BjhihXwfQFEB4FGQe1BzBp+sEFeamrmlg1grmKQnAr03gIo1omIOAYYS45hihr+lQSRG2qRrhgeIeJO7hDuGchR9QCiwRK8jQOm68emfEfQRmxmZQaUJAeUXQdAoax8IAngoYMglATQ9k8AYgKYggReJe2I2AWwEcLermPRXwfR8AAAAkyJwCsQAGyzEKjzGUDmbirjAwCdjkDYArCiSmZiBwgNoDEyBDFoAjFjGuYTFTGKqt4XEwDLGrGcB8iuavG7HsJiAHF9HHHDBpRlBAA), it's used to allow a lower order Modal component to fulfill OBR behaviour even though the UI logic requires a higher order component (Off) to sit above in the vtree.
